[dtensor][experimental] local_map

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123676

**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `local_map` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `local_map` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `redistribute_inputs`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`local_map` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `redistribute_inputs=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

XilunWu committed:
[dtensor][experimental] dmap

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][experimental] dmap"


**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `dmap` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `dmap` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `resharding`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`dmap` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `resharding=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][experimental] dmap"


**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `dmap` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `dmap` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `resharding`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`dmap` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `resharding=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/experimental/dmap.py:
```diff
@@ -0,0 +1,155 @@
+def dmap(
+    func: Callable,
+    out_placements: OutputPlacements,
+    device_mesh: Optional[DeviceMesh] = None,
```
if both `out_placements` and `in_placements` can be None, let's put device_mesh after in_placements

wanchaol commented on torch/distributed/_tensor/experimental/dmap.py:
```diff
@@ -0,0 +1,155 @@
+                    out, DTensor
+                ), f"torch.Tensor output expected but received {type(out)}: {out}"
+
+                flat_dist_out.append(DTensor.from_local(out, device_mesh, spec))
```
from local should call with `run_check=False`, as user mark the out placement sharding we should assume they are correct

wanchaol commented on torch/distributed/_tensor/experimental/dmap.py:
```diff
@@ -0,0 +1,155 @@
+            the device mesh that all the :class:`DTensor`s are placed on. If not specified,
+            this will be inferred from the input :class:`DTensor`s' device mesh. `dmap` requires
+            every :class:`DTensor`s to be placed on the same device mesh.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
```
I think we should explicitly mention that the in_placements are out placements are flattened list if the function's actual inputs/outputs are complicated data structures like nested tuples or dict?

wanchaol commented on torch/distributed/_tensor/experimental/dmap.py:
```diff
@@ -0,0 +1,155 @@
+
+        # we assume every DTensor object is placed on the same device mesh
+        flat_local_args = []
+        nonlocal device_mesh  # access var device_mesh from the outer scope
```
what does this mean? I thought device_mesh from the dmap API should be visible here even without nonlocal?

XilunWu replied:
The reason we need `nonlocal` here is that `device_mesh` may be updated if it's passed in as `None`. Without `nonlocal`, the assignment `device_mesh = arg.device_mesh` will be treated as creating a new local variable `device_mesh` and initializing it to `arg.device_mesh`. If without any `device_mesh = ...` assignment, we would not need the `nonlocal` keyword.

wanchaol reviewed (commented):
first pass, looks pretty good already, please see inlined comments.

tianyu-l reviewed (commented):
Very interesting work! Left some inline comments.

XilunWu commented on torch/distributed/_tensor/experimental/dmap.py:
```diff
@@ -0,0 +1,155 @@
+                    out, DTensor
+                ), f"torch.Tensor output expected but received {type(out)}: {out}"
+
+                flat_dist_out.append(DTensor.from_local(out, device_mesh, spec))
```
Good question! Generally in dtensor we don't check the type of `spec` because in Python without individually checking every element we can only assert if `spec` itself is a `Sequence`. So usually we just check if `spec` is `None` or not.

XilunWu committed:
Update on "[dtensor][experimental] dmap"


**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `dmap` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `dmap` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `resharding`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`dmap` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `resharding=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu commented:
`local_map` should also be able to take in `kwargs` placements as in https://github.com/pytorch/pytorch/pull/124114

XilunWu committed:
Update on "[dtensor][experimental] local_map"


**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `local_map` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `local_map` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `redistribute_inputs`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`local_map` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `redistribute_inputs=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/experimental/test_local_map.py:
```diff
@@ -0,0 +1,200 @@
+def mm_allreduce_forward(device_mesh, W, X):
+    partial_sum_tensor = torch.mm(W, X)
+    reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+    return reduced_tensor
```
nit: maybe explicitly wait for this collective call, otherwise this would leak AsyncColelctiveTensor out

wanchaol commented on test/distributed/_tensor/experimental/test_local_map.py:
```diff
@@ -0,0 +1,200 @@
+    with_comms,
+)
+
+DEVICE_TYPE = "cpu"
```
is there a particular reason to only test this in CPU? Please change this to align with other tests so that we can test in multi gpus

XilunWu replied:
the motivation of limiting it to CPU test is that I think the correctness `local_map` should be consistent on CPU and GPU therefore we only need to test it on CPU.

wanchaol commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+        >>> row_wise = [Shard(0)]  # row-wise sharding placements on 1-d mesh
+        >>> col_wise = [Shard(1)]  # col-wise sharding placements on 1-d mesh
+        >>>
+        >>> # d_mm_allreduce_forward is the function wrapped with DTensor/Tensor convertion
```
nit: we should call this `local_mm_allreduce_forward`

XilunWu replied:
my understanding of the naming is:
`local_map` turns a "local" function into a function on DTensor. The new function should be a "d" function. WDYT?

wanchaol replied:
hmmm isn't it the reverse? 

`local_map` turns a function to a function that escapes DTensor and directly operate on local_tensor?

XilunWu replied:
after offline sync, we decide the naming:
1. `local_map` means this function takes in a `func` that computes on `torch.Tensor` and maps this `func` to its `DTensor` inputs' `_local_tensor`.
2. The return value of `local_map` is a function which 1) performs redistribution on `DTensor` inputs when necessary according to `input_placements`; 2) calls `func` on each `DTensor._local_tensor`; 3) wrap the return value back to `DTensor` by `DTensor.from_local(local_tensor, device_mesh, out_placement)`. This returned function should have prefix `local_` to emphasize that it operates on `DTensor`'s local tensor.

wanchaol commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+        for idx, arg in enumerate(flat_args):
+            if isinstance(arg, DTensor):
+                if device_mesh is None:  # infer device mesh from the DTensor arg
+                    device_mesh = arg.device_mesh
```
Can you add a TODO here? IMO we should probably investigate how this would interact with uneven sharding

wanchaol commented on test/distributed/_tensor/experimental/test_local_map.py:
```diff
@@ -0,0 +1,200 @@
+        )  # row-wisely sharded X tensor
+        # get the function wrapped with DTensor/Tensor convertion
+        # mm_allreduce_forward is a function that applies to Tensors with manual collective
+        # d_mm_allreduce_forward is the function that does the same but applies to DTensors
```
ditto: now these funcs should be called as "local_mm_allreduce_forward"

wanchaol commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor
+from torch.distributed._tensor.placement_types import Placement
+from torch.utils._pytree import tree_flatten, tree_unflatten
```
```
try:
    from torch.utils import _cxx_pytree as pytree
except ImportError:
    from torch.utils import _pytree as pytree  # type: ignore[no-redef]
```
This would make the pytree more efficient

wanchaol commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+        mesh, or if they are placed on a different device mesh than the `device_mesh`
+        argument passed in.
+
+        AssertionError: If the placements in `out_placements` for non-:class:`DTensor`
```
I don't quite get this, could you rephrase this a bit to make it easier to understand?

wanchaol reviewed (approved):
lgtm! have a few comments inlined

yoyoyocmu commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+OutputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def local_map(
```
I think we can have a better function name once the before the release to prod.

XilunWu replied:
any recommendation on the naming? ;-)

yoyoyocmu commented on torch/distributed/_tensor/experimental/local_map.py:
```diff
@@ -0,0 +1,174 @@
+    device_mesh: Optional[DeviceMesh] = None,
+    *,
+    redistribute_inputs: bool = False,
+):
```
I have a use case that in the forward pass, the output of sub-module is of placement: `Replicate()`, but in the backward pass we want the grad's placement to be `_Partial()` to ensure numerical correctness. Wonder in the current API design, are we able to handle it?

XilunWu committed:
Update on "[dtensor][experimental] local_map"


**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `local_map` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `local_map` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `redistribute_inputs`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`local_map` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `redistribute_inputs=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
